dex/pair/src/safe_price_view.rs

```diff
@@ -472,10 +513,76 @@ pub trait SafePriceViewModule:
             second_token_reserve_accumulated,
             weight_accumulated,
             recording_round: search_round,
+            recording_timestamp: interpolated_timestamp,
             lp_supply_accumulated,
         }
     }
 
+    fn find_observation_by_timestamp(
+        &self,
+        target_timestamp: u64,
+        current_index: usize,
+        price_observations: &VecMapper<Self::Api, PriceObservation<Self::Api>, ManagedAddress>,
+    ) -> PriceObservation<Self::Api> {
+        require!(
+            !price_observations.is_empty(),
+            ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+        );
+
+        let last_observation = price_observations.get(current_index);
+        if last_observation.recording_timestamp <= target_timestamp {
+            return last_observation;
+        }
+
+        let mut search_index = 1;
+        let mut left_index;
+        let mut right_index;
+        let observation_at_index_1 = price_observations.get(search_index);
+
+        if observation_at_index_1.recording_timestamp <= target_timestamp {
+            left_index = search_index;
+            right_index = current_index - 1;
+        } else {
+            left_index = current_index + 1;
+            right_index = price_observations.len();
+        }
+
+        let mut closest_observation = observation_at_index_1.clone();
+        let mut min_timestamp_diff =
+            if target_timestamp > observation_at_index_1.recording_timestamp {
+                target_timestamp - observation_at_index_1.recording_timestamp
+            } else {
+                observation_at_index_1.recording_timestamp - target_timestamp
+            };
+
+        while left_index <= right_index {
+            search_index = (left_index + right_index) / 2;
+            let current_observation = price_observations.get(search_index);
+            let current_timestamp_diff =
+                if target_timestamp > current_observation.recording_timestamp {
+                    target_timestamp - current_observation.recording_timestamp
+                } else {
+                    current_observation.recording_timestamp - target_timestamp
+                };
+
+            if current_timestamp_diff < min_timestamp_diff {
+                min_timestamp_diff = current_timestamp_diff;
+                closest_observation = current_observation.clone();
+            }
```

Comment:
Can be its own function.

Reply:
Well, I personally think that making code easier to read is the most important thing :)

Reply:
This doesn't work unless we pass around a lot of parameters, so I would avoid the extra complications. I've tried to do it as minimal as possible, but not succeeded.